gsplat/strategy.py

```diff
@@ -0,0 +1,552 @@
+from abc import ABC, abstractmethod
+from collections import defaultdict
+from typing import Any, Callable, DefaultDict, Dict, List, Tuple
+
+import torch
+import torch.nn.functional as F
+from torch import Tensor
+
+from gsplat import quat_scale_to_covar_preci
+from gsplat.relocation import compute_relocation
+from gsplat.utils import normalized_quat_to_rotmat
+
+
+class Strategy(ABC):
+    """Base class for the GS optimization strategy."""
+
+    def __init__(
+        self,
+        params: torch.nn.ParameterDict,
+        optimizers: List[torch.optim.Optimizer],
+        *args,
+        **kwargs,
+    ):
+        self._params = params
+        self._optimizers = optimizers
+        # some running states for the strategy
+        self._state: DefaultDict[str, Any] = defaultdict(dict)
+
+    @abstractmethod
+    def step_pre_backward(self, step: int, info: Dict[str, Any]):
+        raise NotImplementedError
+
+    @abstractmethod
+    def step_post_backward(self, step: int, info: Dict[str, Any]):
+        raise NotImplementedError
+
+
+class NullStrategy(Strategy):
+    """A null strategy that does nothing."""
+
+    def step_pre_backward(self, step: int, info: Dict[str, Any]):
+        pass
+
+    def step_post_backward(self, step: int, info: Dict[str, Any]):
+        pass
+
+
+class DefaultStrategy(Strategy):
+    """A default strategy that follows the original 3DGS paper:
+
+    "3D Gaussian Splatting for Real-Time Radiance Field Rendering"
+    https://arxiv.org/abs/2308.04079
+
+    If absgrad=True, it will use the absolute gradients for GS splitting, following the
+    AbsGS paper:
+
+    "AbsGS: Recovering Fine Details for 3D Gaussian Splatting"
+    https://arxiv.org/abs/2404.10484
+    """
+
+    def __init__(
+        self,
+        params: torch.nn.ParameterDict,
+        optimizers: List[torch.optim.Optimizer],
+        verbose: bool = False,
+        scene_scale: float = 1.0,
+        # GSs with opacity below this value will be pruned
+        prune_opa: float = 0.005,
+        # GSs with image plane gradient above this value will be split/duplicated
+        grow_grad2d: float = 0.0002,
+        # GSs with scale below this value will be duplicated. Above will be split
+        grow_scale3d: float = 0.01,
+        # GSs with scale above this value will be pruned.
+        prune_scale3d: float = 0.1,
+        # Start refining GSs after this iteration
+        refine_start_iter: int = 500,
+        # Stop refining GSs after this iteration
+        refine_stop_iter: int = 15_000,
+        # Reset opacities every this steps
+        reset_every: int = 3000,
+        # Refine GSs every this steps
+        refine_every: int = 100,
+        # Use absolute gradients for GS splitting
+        absgrad: bool = False,
```

Comment:
if this is true maybe we want to automatically set the grow_grad2d to .008 otherwise the number of gaussians will explode? or maybe warn the user?

Reply:
Not sure if warning is a good idea. on different scenes this threshold could be very different. Auto set is also not ideal as we want to give the user the freedom to adjust that. 

I think the best we can do is probably note it clearly in the docstring.

Reply:
Or, another idea is to create another class `AbsGradStrategy` that inherit from this class and set the default value to 0.008.